### ISSUE TITLE
Cannot write dask Dataset to NetCDF file

### ISSUE THREAD

**Issue by Scheibs** [pydata/xarray#729]
I have a xarray Dataset created using dask, which i would like to write on my disk.
The size of this dataset is 12GB, and my computer has 64GB of memory, but when I run the 'to_netcdf' command, my memory crashes.
```python
type(nc)
Out[5]: xray.core.dataset.Dataset

nc.chunks
Out[6]: Frozen(SortedKeysDict({'Denree': (19,), u'NIsoSource': (10, 10, 10, 10, 10, 6), 'Pop': (6,),
u'DimJ0': (20, 17), u'DimK0': (1,), u'time': (50, 50, 50, 50, 3), u'DimI0': (15,), 'TypeDose': (2,)}))

nc.nbytes * (2 ** -30)
Out[7]: 12.4569926
```
I don't understand what i'm doing wrong, so thanks for your help.

**Comment by shoyer** [comment]
This is almost certainly an issue with dask's scheduler. cc @mrocklin

Could you share a summary of how you create this dataset? Is it something that should be possible to calculate in a single pass over the data?

**Comment by mrocklin** [comment]
@Scheibs can you try calling these lines to remove multi-threading and see if the problem persists?

```python
import dask
dask.set_options(get=dask.async.get_sync)
```

I agree with @shoyer that it would be very useful to see what you're doing that causes this problem.

**Comment by Scheibs** [comment]
@mrocklin I have tried your lines, it seemed to work but finally I got the memory error message. However the crash was different than what I used to have...

My dataset is a set of several variables, which are calculated like this :

I have DataArray of unitary data for four rainfall height(0,5,10,15 mm/days,"SYMB") and another DataArray with spatialized rainfall height values ("RAIN").

```
SYMB <xray.DataArray (Hpluie: 4, time: 203, NIsoSource: 56, Denree: 19, Paliers: 23)>
dask.array<xray-Fo..., shape=(4, 203, 56, 19, 23), dtype=float64, chunksize=(4, 50, 10, 19, 5)>
```

```
RAIN <xray.DataArray (Paliers: 23, DimK0: 1, DimJ0: 37, DimI0: 15)>
dask.array<xray-Ra..., shape=(23, 1, 37, 15), dtype=float64, chunksize=(5, 1, 20, 15)>
```

Then I created an array which is the result of interpolation between this two arrays
The function  is :
```python
def interp(array,pluie):

    p0 = (array.sel(Hpluie = 5) -array.sel(Hpluie = 0))/5
    p1 = (array.sel(Hpluie = 10) -array.sel(Hpluie = 5))/5
    p2 = (array.sel(Hpluie = 15) -array.sel(Hpluie = 10))/5

    interp = (p0*(pluie.where(pluie<5) - array.Hpluie.sel(Hpluie=5)) +array.sel(Hpluie=5)).fillna(0)\
        +(p1*(pluie.where((pluie>=5) & (pluie<=10)) - array.Hpluie.sel(Hpluie=5)) +array.sel(Hpluie=5)).fillna(0)\
        +(p2*(pluie.where(pluie>10) - array.Hpluie.sel(Hpluie=10)) + array.sel(Hpluie=10)).fillna(0)   
    return interp
```

So  M = interp(SYMB,RAIN)
```
M  <xray.DataArray (time: 203, NIsoSource: 56, Denree: 19, Paliers: 23, DimK0: 1, DimJ0: 37, DimI0: 15)>
dask.array<elemwis..., shape=(203, 56, 19, 23, 1, 37, 15), dtype=float64, chunksize=(50, 10, 19, 5, 1, 20, 15)>
```

My final variable is the product of this interpolated array and an other spatialized variable("FLUX")

```
FLUX <xray.DataArray 'Flux_Moyen_Depot_Sec' (Paliers: 23, DimK0: 1, DimJ0: 37, DimI0: 15, NIsoSource: 56)>
dask.array<xray-Fl..., shape=(23, 1, 37, 15, 56), dtype=float64, chunksize=(5, 1, 20, 15, 10)>
```

RES = (M * FLUX).sum("Paliers")

```
RES = <xray.DataArray (DimK0: 1, DimJ0: 37, DimI0: 15, NIsoSource: 56, time: 203, Denree: 19)>
dask.array<p_reduc..., shape=(1, 37, 15, 56, 203, 19), dtype=float64, chunksize=(1, 20, 15, 10, 50, 19
```
I hope this is at less understandable... 
thanks a lot for your help

**Comment by mrocklin** [comment]
Sorry for the delay in response.  

Nothing here seems dangerous to me.  @shoyer does the writeup above raise any questions for you?

If convenient, it would be interesting to see the output of a few of the dask profilers:

```
$ conda install bokeh
$ pip install cachey
```

```python
from dask.diagnostics import CacheProfiler, ResourceProfiler, Profiler, visualize
with Profiler() as prof, CacheProfiler(metric=cachey.nbytes) as cprof, ResourceProfiler() as rprof:
    # call the final dataset.to_netcdf() function

visualize([prof, cprof, rprof], file_path='profile.html')
```

And then upload that file somewhere, perhaps to a [gist](https://gist.github.com/).  In order to make this run to completion you might have to operate on a subset of the dataset.

Alternatively, is there a way for me to recreate a version of this dataset on my local machine?  @shoyer is there a way to capture the metadata of netcdf files and reinstantiate empty copies of them on another machine?

**Comment by Scheibs** [comment]
I can send you my files, with the two netCDF4 source files and the script. I will also join the Bokeh graph for only one variable which fit with memory.
Do you know how I can send you the ZIP folder ? It's 830 MB.

**Comment by shoyer** [comment]
Something like `einsum` or a broadcasting matrix multiplication could help a little bit here, by replacing `(M * FLUX).sum('Paliers')` with `M.dot(FLUX, dim='Paliers')` and thereby reducing peak memory consumption, but even there I'm calculating peak chunk size at 950000 elements. This should be totally fine on most machines.

@mrocklin Unfortunately, I don't know an easy way to create a copy of a netCDF file with random data, but that's a good idea for a little project....

**Comment by mrocklin** [comment]
I would generally send such a large file by hosting it at a web-accessible location.  Perhaps you are at an institution were you have access to host files online?

**Comment by Scheibs** [comment]
@mrocklin I put my folder on a ftp server, can I have an email address to send you the login information ?

@shoyer I have tried replacing sum() by dot() but I get an error "Data Array has no attribute 'dot' "

**Comment by mrocklin** [comment]
mrocklin continuum io

**Comment by Scheibs** [comment]
@mrocklin I finally gave you a free acces path on ftp://ftp.irsn.fr/argon/SHARE/

**Comment by mrocklin** [comment]
My apologies for the slow response (very busy week, lots of exciting stuff, sadly results in poor user response). 

I can access that page easily but the download seems to halt after 11MB

**Comment by Scheibs** [comment]
You don't have to apologize for the delay, i'm already grateful for your time on this !

I don't understand why this download won't work with you; I tried with other and everything was ok...
Maybe you can try with this link  :  
[https://sharing.oodrive.com/easyshare/fwd/link=BS6TALAs9CiQg.WNotfITA](https://sharing.oodrive.com/easyshare/fwd/link=BS6TALAs9CiQg.WNotfITA)

**Comment by Scheibs** [comment]
@mrocklin is that new link working ? Because it's gonna be closed tomorow.

**Comment by mrocklin** [comment]
Downloaded

**Comment by mrocklin** [comment]
Slowly taking a look at this now.  Large PDF for the full computation, if anyone is interested:

[dask.pdf](https://github.com/pydata/xarray/files/131262/dask.pdf)

**Comment by mrocklin** [comment]
Looking at the task graph my first guess is that @shoyer is correct, and that we've found another case that the scheduler should be able to handle well, but doesn't.  This hasn't happened in a while, but it always leads to improvements whenever we find such a problem.

For a case this complex I think we either need to reduce it to a particular graph motif on which we schedule poorly or we first need to develop a better way to visualize traces of the scheduler's behavior.  I've started a separate dask issue: https://github.com/dask/dask/issues/994

For the near future I don't have a solution to @Scheibs 's research problem (sorry!)  This will probably require tweaking dask scheduler internals which probably won't happen by me in the next couple of weeks.  I'm very happy that people brought this to my attention though.
